src/util/locktime.rs

```diff
@@ -0,0 +1,516 @@
+    /// assert!(locktime::is_block_height(n_lock_time));
+    /// let lock_time = LockTime::from(n_lock_time);
+    /// assert_eq!(lock_time.to_u32(), n_lock_time);
+    pub fn to_u32(&self) -> u32 {
+        self.0
+    }
+}
+
+impl TryFrom<u32> for Height {
+    type Error = Error;
+
+    /// Constructs a new block height.
+    ///
+    /// # Errors
+    /// If `n` does not represent a block height value (see [`is_block_height`]).
+    fn try_from(n: u32) -> Result<Height, Self::Error> {
+        if is_block_height(n) {
+            Ok(Self(n))
+        } else {
+            Err(ConversionError::invalid_height(n).into())
+        }
+    }
+}
+
+impl From<Height> for u32 {
+    fn from(h: Height) -> Self {
+        h.0
+    }
+}
+
+impl fmt::Display for Height {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+/// A UNIX timestamp, seconds since epoch.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct Time(u32); // Guaranteed to always contain a valid time value.
+
+impl Time {
+    /// Converts this `Time` to its inner `u32` value.
+    ///
+    /// # Examples
+    /// ```rust
+    /// use bitcoin::locktime::{self, LockTime};
+    ///
+    /// let n_lock_time: u32 = 1653195600; // May 22nd, 5am UTC.
+    /// assert!(locktime::is_block_time(n_lock_time));
+    /// let lock_time = LockTime::from(n_lock_time);
+    /// assert_eq!(lock_time.to_u32(), n_lock_time);
+    /// ```
+    pub fn to_u32(&self) -> u32 {
+        self.0
+    }
+}
+
+impl TryFrom<u32> for Time {
+    type Error = Error;
+
+    /// Constructs a new block time.
+    ///
+    /// # Errors
+    /// If `n` does not encode a UNIX time stamp (see [`is_block_time`]).
+    fn try_from(n: u32) -> Result<Time, Self::Error> {
+        if is_block_time(n) {
+            Ok(Self(n))
+        } else {
+            Err(ConversionError::invalid_time(n).into())
+        }
+    }
+}
+
+impl From<Time> for u32 {
+    fn from(t: Time) -> Self {
+        t.0
+    }
+}
+
+impl fmt::Display for Time {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+/// Catchall type for errors that relate to timelocks.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[non_exhaustive]
+pub enum Error {
+    /// An error occurred while converting a `u32` to a lock time variant.
+    Conversion(ConversionError),
+    /// An error occurred while operating on lock times.
+    Operation(OperationError),
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match *self {
+            Conversion(ref e) => write_err!(f, "error converting lock time value"; e),
+            Operation(ref e) => write_err!(f, "error during lock time operation"; e),
+        }
+    }
+}
```

Comment:
Not required but `FromStr`, `TryFrom<&str>`, `TryFrom<String>`... could be nice. Too. We would have to add `ParseError` type though.

Reply:
Better to just return the `core::num::ParseIntError` directly, right?
```rust
impl FromStr for LockTime {
    type Err = ParseIntError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let n: u32 = s.parse()?;
        Ok(LockTime::from(n))
    }
}
```

Reply:
Added as separate PR so it can be seen easily, its patch 2, is that what you wanted?

Reply:
Hmm, yes, `ParseIntError` looks OK, though it could be more tricky to `?` it with other errors we have. Also I realized there's implicit consensus representation which I'm against, so not sure if the whole thing is a good idea. :D